0000-friends_in_high_kindednesses.md

```diff
@@ -0,0 +1,442 @@
+syntax is that many users will be able to use types which have assocaited type
+constructors without even being aware that this has something to do with a type
+system feature called higher-kindedness.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Adding language complexity
+
+This would add a somewhat complex feature to the language, being able to
+polymorphically resolve type constructors, and requires several extensions to
+the type system which make the implementation more complicated.
+
+Additionally, though the syntax is designed to make this feature easy to learn,
+it also makes it more plausible that a user may accidentally use it when they
+mean something else, similar to the confusion between `impl .. for Trait` and
+`impl<T> .. for T where T: Trait`. For example:
+
+```rust
+// The user means this
+trait Foo<'a> {
+    type Bar: 'a;
+}
+
+// But they write this
+trait Foo<'a> {
+    type Bar<'a>;
+}
+```
+
+## Not full "higher-kinded types"
+
+This does not add all of the features people want when they talk about higher-
+kinded types. For example, it does not enable traits like `Monad`. Some people
+may prefer to implement all of these features together at once. However, this
+feature is forward compatible with other kinds of higher-kinded polymorphism,
+and doesn't preclude implementing them in any way. In fact, it paves the way
+by solving some implementation details that will impact other kinds of higher-
+kindedness as well, such as partial application.
+
+## Syntax isn't like other forms of higher-kinded polymorphism
+
+Though the proposed syntax is very similar to the syntax for associated types
+and type aliases, it is probably not possible for other forms of higher-kinded
+polymorphism to use a syntax along the same lines. For this reason, the syntax
+used to define an associated type constructor will probably be very different
+from the syntax used to e.g. implement a trait for a type constructor.
+
+However, the syntax used for these other forms of higher-kinded polymorphism
+will depend on exactly what features they enable. It would be hard to design
+a syntax which is consistent with unknown features.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Push HRTBs harder without associated type constructors
+
+An alternative is to push harder on HRTBs, possibly introducing some elision
+that would make them easier to use.
+
+Currently, an approximation of `StreamingIterator` can be defined like this:
+
+```rust
+trait StreamingIterator<'a> {
+   type Item: 'a;
+   fn next(&'a self) -> Option<Self::Item>;
+}
+```
+
+You can then bound types as `T: for<'a> StreamingIterator<'a>` to avoid the
+lifetime parameter infecting everything `StreamingIterator` appears.
+
+However, this only partially prevents the infectiveness of `StreamingIterator`,
+only allows for some of the types that associated type constructors can
+express, and is in generally a hacky attempt to work around the limitation
```

Comment:
Firstly, none of these traits can be expressed without this feature as they are. Something similar can be expressed by pushing the parameter upward but its not the _same_ trait.

I think when I wrote this I was thinking about ATCs with type parameters but that's no longer in the RFC.